README.md
xsos & rsar - Summarize system info from sosreports
===================================================

**Jump to ...**

* [I'M LOOKING FOR RSAR](/RSAR_README.md)
* [CHANGELOG](/CHANGELOG.md)
* [SCREEN SHOTS](#screen-shots)
* [INTRO](#intro)
* [INSTALLATION](#installation)
* [EXAMPLES IN ACTION](#examples-in-action)
* [REQUIREMENTS](#requirements)
* [THINGS THAT MIGHT SURPRISE YOU](#things-that-might-surprise-you)
* [AUTHORS](#authors)
* [LICENSE](#license)


SCREEN SHOTS
-------

**`xsos -om` OR `xsos --os --mem` would show something like the following:**

![xsos parsing general os info and /proc/meminfo](http://b19.org/linux/xsos-osmem.png)

**`xsos -cln` OR `xsos --cpu --lspci --net` could show this:**

![xsos parsing /proc/cpuinfo, lspci, /proc/net/dev](http://b19.org/linux/xsos-cln.png)

**`xsos -e` OR `xsos --ethtool` output looks like this:**

![xsos parsing ethtool output](http://b19.org/linux/xsos-e.png)


INTRO
-------

The goal of `xsos` is to make it easy to instantaneously gather information about a system together in an easy-to-read-summary, whether that system is the localhost on which xsos is being run or a system for which you have an unpacked sosreport.

There is tons of useful amazing instructive data available to normal users (not to mention root) on a Linux system, but by design, this information is spread out across myriad files. Some of it even requires commands to parse through.

`xsos` will attempt to make it easy, parsing and calculating and formatting data from dozens of files (and commands) to give you a detailed overview about a system, or -- if requested -- `xsos` will only parse one file (e.g. with `--mem` or `--cpu`) or output from one command (e.g. with `--ip` or `--ps`).

New features are being added all the time -- see [the tracker](https://github.com/ryran/xsos/issues) to have a look at some of the things that are already in line to be worked on.

**[Why another tool? Why not add stuff to `sxconsole`?](https://github.com/ryran/xsos/issues/9)**


INSTALLATION
-------

*"I'm sold! How do I install xsos?"*

Two choices:

1. **Manual install:**
   Get the very latest (potentially bleeding-edge) version directly by downloading the main xsos file from [bit.ly/xsos-direct](http://bit.ly/xsos-direct) (which points to [raw.github.com/ryran/xsos/master/xsos](https://raw.github.com/ryran/xsos/master/xsos)). Updating will need to be done manually by re-downloading.
   Explicitly, you could run the following as root:
   
    ```
    wget -O /usr/local/bin/xsos bit.ly/xsos-direct
    chmod +x /usr/local/bin/xsos
    ```

2. **RPM/Yum repo (recommended):**
   Configure access to the Fedora/RHEL6+ yum repository @ [people.redhat.com/rsawhill/rpms](http://people.redhat.com/rsawhill/rpms/) and install the [potentially not-as-bleeding-edge] xsos rpm from there. Updating can happen automatically along with the rest of your system.
   Two-step instructions @ [How to install xsos or rsar via yum?](https://github.com/ryran/xsos/issues/67#issuecomment-23339115)

Once xsos is installed by one of the above methods, run `xsos -h` as a normal user to see the help page and get started.


EXAMPLES IN ACTION
-------
The lovely thing that isn't going to be captured here is all the coloring done to make things easier to read. There's actually quite a bit of color-logic to warn of various conditions...

**Run on a sosreport with no options:**

```
[rsaw]$ xsos aczx998pinkle/
OS
  Hostname: aczx998pinkle
  Distro:   [redhat-release] Red Hat Enterprise Linux Server release 5.5 (Carthage)
            [enterprise-release] Enterprise Linux Enterprise Linux Server release 5.5 (Carthage)
  RHN:      serverURL=https://linux-update.oracle.com/XMLRPC
  Runlevel: N 3  (default 3)
  SELinux:  permissive  (default enforcing)
  Arch:     mach=x86_64  cpu=x86_64  platform=x86_64
  Kernel:
    Booted kernel:  2.6.18-238.12.2.0.2.el5
    GRUB default:   2.6.18-238.12.2.0.2.el5
    Build version:
      Linux version 2.6.18-238.12.2.0.2.el5 (mockbuild@ca-build9.us.oracle.com) (gcc version 4.1.2 
      20080704 (Red Hat 4.1.2-50)) #1 SMP Tue Jun 28 05:21:19 EDT 2011
    Booted kernel cmdline:
      root=/dev/md6 ro bootarea=c0d0 loglevel=7 panic=60 debug rhgb numa=off console=ttyS0,115200n8 
      console=tty1 crashkernel=128M@16M bootfrom=CELLBOOT audit=1 processor.max_cstate=1 nomce
    GRUB default kernel cmdline:
      root=/dev/md6 ro bootarea=c0d0 loglevel=7 panic=60 debug rhgb numa=off console=ttyS0,115200n8 
      console=tty1 crashkernel=128M@16M bootfrom=BOOT audit=1 processor.max_cstate=1 nomce
    Kernel taint-check: 536870912 512 64 16 1
      Technology Preview code is loaded
      Taint on warning
      Userspace-defined naughtiness
      System experienced a machine check exception
      Proprietary module has been loaded
    - - - - - - - - - - - - - - - - - - -
  Sys time:  Mon Oct 29 10:55:02 CDT 2012
  Boot time: Sat Apr 28 03:29:56 CDT 2012  (1335583796)
  Uptime:    184 days, 12:25,  1 user
  LoadAvg:   [24 CPU] 2.34 (10%), 1.27 (5%), 0.95 (4%)
  /proc/stat:
    procs_running: 4    procs_blocked: 1    processes: 248052571
    cpu:  [Break-down of CPU time since boot]
      us 1%, ni 0%, sys 1%, idle 96%, iowait 2%, irq 0%, sftirq 0%, steal 0%
```


**While xsos is always being improved, here's the minimal help page from v0.5.6:**

```
Usage: xsos [DISPLAY OPTIONS] [-6abokcmdtlerngisp] [SOSREPORT ROOT]
  or:  xsos [DISPLAY OPTIONS] {--B|--C|--M|--D|--T|--L|--R|--N|--G|--I|--P FILE}...
  or:  xsos [-?|-h|--help]
  or:  xsos [-U|--update]

Display system info from localhost or extracted sosreport

Content options:
 -a, --all      show everything
 -b, --bios     show info from dmidecode
 -o, --os       show hostname, distro, SELinux, kernel info, uptime, etc
 -k, --kdump    inspect kdump configuration
 -c, --cpu      show info from /proc/cpuinfo
 -m, --mem      show info from /proc/meminfo
 -d, --disks    show info from /proc/partitions + dm-multipath synopsis
 -t, --mpath    show info from dm-multipath
 -l, --lspci    show info from lspci
 -e, --ethtool  show info from ethtool
 -r, --softirq  show info from /proc/net/softnet_stat
 -n, --netdev   show info from /proc/net/dev
 -g, --bonding  show info from /proc/net/bonding
 -i, --ip       show info from ip addr (BASH v4+ required)
     --net      alias for: --lspci --ethtool --softirq --netdev --bonding --ip
 -s, --sysctl   show important kernel sysctls
 -p, --ps       inspect running processes via ps

Display options:
     --scrub-ip     remove IP addresses & hostnames from output
     --scrub-mac    remove HW MAC addresses from output
                    see XSOS_SCRUB_IP_HN & XSOS_SCRUB_MACADDR env vars
 -6, --ipv6         parse ip addr output for IPv6 addresses instead of IPv4
 -q, --wwid=ID      restrict dm-multipath output to a particular mpath device,
                    where ID is a wwid, friendly name, or LUN identifier
 -u, --unit=P       change byte display for /proc/meminfo & /proc/net/dev,
                    where P is "b" for byte, or else "k", "m", "g", or "t"
 -v, --verbose=NUM  specify ps verbosity level (0-4, default: 1)
 -w, --width=NUM    change fold-width, in columns (positive number, e.g., 80)
                    "0" disables wrapping, "w" autodetects width (default)
 -x, --nocolor      disable output colorization
 -y, --less         send output to `less -SR`
 -z, --more         send output to `more`

Special options (BASH v4+ required):
 --B=FILE  read from FILE containing `dmidecode` dump
 --C=FILE  read from FILE containing /proc/cpuinfo dump
 --M=FILE  read from FILE containing /proc/meminfo dump
 --D=FILE  read from FILE containing /proc/partitions dump
 --T=FILE  read from FILE containing `multipath -v4 -ll` dump
 --L=FILE  read from FILE containing `lspci` dump
 --R=FILE  read from FILE containing /proc/net/softnet_stat dump
 --N=FILE  read from FILE containing /proc/net/dev dump
 --G=FILE  read from FILE containing /proc/net/bonding/xxx dump
 --I=FILE  read from FILE containing `ip addr` dump
 --P=FILE  read from FILE containing `ps aux` dump

Run with "--help" to see full help page

Version info: xsos v0.5.6 last mod 2015/01/01
See <github.com/ryran/xsos> to report bugs or suggestions
```


**Here it is run on my laptop as root, with specific options:**

```
[rsaw]$ sudo xsos --bios --ip --lspci --ethtool --net --ps  # would be same as 'xsos -bilenp'
DMIDECODE
  BIOS:
    LENOVO, version 8CET50WW (1.30 ), 11/01/2011
  System:
    Mfr:  LENOVO
    Prod: 4174AQ5
    Vers: ThinkPad T420s
    Ser:  R9XXXXXXXX
    UUID: 56EABD01-XXXX-11CB-9DE6-XXXXXXXXXX
  CPU:
    1 of 1 CPU sockets populated, 2 cores/4 threads per CPU
    2 total cores, 4 total threads
    Mfr:  Intel(R) Corporation
    Fam:  Core i5
    Freq: 2600 MHz
    Vers: Intel(R) Core(TM) i5-2540M CPU @ 2.60GHz
  Memory:
    8192 MB (8 GB) total
    2 of 4 DIMMs populated (system max capacity 16 GB)

LSPCI
  Net/Storage:
    (1) Intel Corporation Centrino Ultimate-N 6300 (rev 3e)
    (1) Intel Corporation 82579LM Gigabit Network Connection (rev 04)
  VGA:
    Intel Corporation 2nd Generation Core Processor Family Integrated Graphics Controller (rev 09)

ETHTOOL
  Interface Status:
    em1         link=up 1000Mb/s full (autoneg=Y)  drv e1000e v2.2.14-k / fw 0.13-3
    tun0        link=up 10Mb/s full (autoneg=N)    drv tun v1.6
    virbr0      link=DOWN                          drv bridge v2.3 / fw N/A
    virbr0-nic  link=DOWN                          drv tun v1.6
    wlan0       link=up                            drv iwlwifi v3.9.6-200.fc18.x86_64 / fw 9.221.4.1
  Interface Errors:
    em1  dropped_smbus: 267


IP
  Interface   Slave Of  IPv4 Address        State  MAC Address
  =========   ========  ==================  =====  =================
  lo          -         127.0.0.1/8         up     -
  em1         -         192.168.0.20/24     up     f0:de:f1:ba:cf:c4
  wlan0       -         10.7.7.111/24       up     24:77:03:41:41:18
                        192.168.17.17/32            
  virbr0      -         192.168.122.1/24    up     52:54:00:97:e3:ee
  virbr0-nic  -         -                   DOWN   52:54:00:97:e3:ee
  tun0        -         10.10.49.81/32      up     -

NETDEV
  Interface   RxMiBytes  RxPackets  RxErrs  RxDrop  TxMiBytes  TxPackets  TxErrs  TxDrop  TxColls
  =========   =========  =========  ======  ======  =========  =========  ======  ======  =======
  em1         6715       6940 k     0       0       6715       1972 k     0       0       0 
  tun0        0          0 k        0       0       0          0 k        0       0       0 
  virbr0      8          41 k       0       0       8          37 k       0       0       0 
  virbr0-nic  0          0 k        0       0       0          0 k        0       0       0 
  wlan0       1902       1761 k     0       0       1902       1139 k     0       0       0 


PS CHECK
  Total number of processes: 
    291
  Top users of CPU & MEM: 
    USER  %CPU   %MEM   RSS 
    rsaw  34.1%  46.0%  3.81 GiB
    qemu  27.4%  4.9%   0.38 GiB
    root  2.7%   1.0%   0.20 GiB
  Uninteruptible sleep & Defunct processes: 
    USER     PID    %CPU  %MEM  VSZ-MiB  RSS-MiB  TTY    STAT   START  TIME    COMMAND
    rsaw     5287   0.0   0.0   0        0        pts/0  D+     17:19  0:00    [cat]
    rsaw     5125   0.0   0.0   0        0        ?      Z      Dec16  0:48    [chromium-browse] <defunct> 
    rsaw     8307   0.2   0.0   0        0        ?      Z      Dec17  3:07    [chromium-browse] <defunct> 
    rsaw     8867   0.0   0.0   0        0        ?      Z      Dec17  1:20    [xchat] <defunct> 
    rsaw     9533   0.0   0.0   0        0        ?      Z      Dec17  0:32    [gedit] <defunct> 
  Top CPU-using processes: 
    USER     PID    %CPU  %MEM  VSZ-MiB  RSS-MiB  TTY    STAT   START  TIME    COMMAND
    qemu     613    27.4  4.9   3094     393      ?      Sl     00:38  0:29    /usr/bin/qemu-kvm -S -M 
    rsaw     23059  13.7  2.3   1091     185      ?      Sl     Dec26  21:34   /proc/self/exe --type=plugin --plugin-path=/usr/lib64/flash-plugin/libflashplayer.so 
    rsaw     621    5.9   0.8   964      65       ?      Ss     00:38  0:06    python /usr/share/virt-manager/virt-manager.py 
    rsaw     2137   5.3   18.7  2705     1472     ?      Sl     Dec24  204:37  /usr/lib64/firefox/firefox 
    root     1226   2.2   0.5   135      42       tty1   Ss+    Dec24  87:57   /usr/bin/Xorg :0 -background 
    rsaw     11901  2.1   4.2   2319     335      ?      SLl    Dec24  74:54   /usr/bin/gnome-shell 
    rsaw     22866  2.0   1.6   898      128      ?      Sl     Dec26  3:11    /usr/lib64/chromium-browser/chromium-browser --enable-plugins --enable-extensions 
    rsaw     29408  1.3   0.5   1218     39       ?      Sl     Dec26  0:37    gedit 
    rsaw     23084  0.8   1.2   1140     99       ?      Sl     Dec26  1:18    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
    rsaw     22952  0.8   1.7   1168     141      ?      Sl     Dec26  1:22    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
  Top MEM-using processes: 
    USER     PID    %CPU  %MEM  VSZ-MiB  RSS-MiB  TTY    STAT   START  TIME    COMMAND
    rsaw     2137   5.3   18.7  2705     1472     ?      Sl     Dec24  204:37  /usr/lib64/firefox/firefox 
    qemu     613    27.4  4.9   3094     393      ?      Sl     00:38  0:29    /usr/bin/qemu-kvm -S -M 
    rsaw     11901  2.1   4.2   2319     335      ?      SLl    Dec24  74:54   /usr/bin/gnome-shell 
    rsaw     23059  13.7  2.3   1091     185      ?      Sl     Dec26  21:34   /proc/self/exe --type=plugin --plugin-path=/usr/lib64/flash-plugin/libflashplayer.so 
    rsaw     22952  0.8   1.7   1168     141      ?      Sl     Dec26  1:22    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
    rsaw     22866  2.0   1.6   898      128      ?      Sl     Dec26  3:11    /usr/lib64/chromium-browser/chromium-browser --enable-plugins --enable-extensions 
    rsaw     23214  0.6   1.5   1141     121      ?      Sl     Dec26  0:59    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
    rsaw     23004  0.0   1.2   1130     100      ?      Sl     Dec26  0:06    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
    rsaw     22959  0.0   1.2   1127     100      ?      Sl     Dec26  0:05    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
    rsaw     23084  0.8   1.2   1140     99       ?      Sl     Dec26  1:18    /usr/lib64/chromium-browser/chromium-browser --type=renderer --lang=en-US 
```



**Run on another sosreport with some options:**

```
[rsaw]$ xsos 8308201prodserv --disks --mem --unit=m
MEMORY
  Stats graphed as percent of MemTotal:
    MemUsed    ◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◇◇  96.0%
    HugePages  ◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◆◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇  60.5%
    Buffers    ◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇   0.6%
    Cached     ◆◆◆◆◆◆◆◆◆◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇  18.6%
    Dirty      ◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇◇   0.0%
  RAM:
    64363 MiB total ram
    61815 MiB (96%) used
    49451 MiB (76.8%) used excluding Buffers/Cached
    5 MiB (0%) dirty
  HugePages:
    38912 MiB pre-allocated to HugePages (60.5% of total ram)
    38146 MiB of HugePages (98%) in-use by applications
  LowMem/Slab/PageTables/Shmem:
    61815 MiB (96%) of 64363 MiB LowMem in-use
    1192 MiB (1.9%) of total ram used for Slab
    349 MiB (0.5%) of total ram used for PageTables
  Swap:
    464 MiB (2.8%) used of 16381 MiB total

STORAGE
  Multipath:
    archive    2.0 T
    dbbackup   850 G
    dgspool    20 G
    miscspool  100 G
    proddb     200 G
    prodstage  100 G
    produpgdb  150 G
    reportdb   30 G
    tsttrndev  150 G
  Disks:
    (Multipath and/or software raid components hidden)
    1 disks, totaling 272 GiB (0.27 TiB)
    sda   271.9 G

[rsaw]$ xsos --net --unit k 8308201prodserv
NETDEV
  Interface  RxKiBytes     RxPackets     RxErrs  RxDrop     TxKiBytes     TxPackets     TxErrs  TxDrop
  =========  =========     =========     ======  ======     =========     =========     ======  ======
  eth0       23569277      300080406     0       0          23569277      480926218     0       0 
  eth1       144484269105  342207919183  0       4494 (0%)  144484269105  357978804158  0       0 
  eth2       428438229206  392560553366  0       373 (0%)   428438229206  330736051661  0       0 
  eth4       37184228554   55232280858   0       2350 (0%)  37184228554   35535848225   0       0 


[rsaw]$ xsos -nug 8308201prodserv
NETDEV
  Interface  RxGiBytes  RxPackets  RxErrs  RxDrop     TxGiBytes  TxPackets  TxErrs  TxDrop
  =========  =========  =========  ======  ======     =========  =========  ======  ======
  eth0       22.5       300 M      0       0          22.5       481 M      0       0 
  eth1       137791     342208 M   0       4494 (0%)  137791     357979 M   0       0 
  eth2       408590     392561 M   0       373 (0%)   408590     330736 M   0       0 
  eth4       35461.6    55232 M    0       2350 (0%)  35461.6    35536 M    0       0 
```



**Run on a sosreport again, showing sysctls:**

```
[rsaw]$ xsos --sysctl boomyaow/
SYSCTLS
  kernel.
    hostname =  boomyaow03
    osrelease =  2.6.32-220.13.1.el6.x86_64
    tainted =  0  (kernel untainted)
    random.boot_id =  defb55c5-c2bb-4e4e-86c3-745d80a6585d
    random.entropy_avail [bits] =  154
    hung_task_panic [bool] =  0
    hung_task_timeout_secs =  120  (secs task must be D-state to trigger)
    msgmax [bytes] =  65536
    msgmnb [bytes] =  65536
    msgmni [msg queues] =  32768
    panic [secs] =  0  (no autoreboot on panic)
    panic_on_oops [bool] =  1
    panic_on_unrecovered_nmi [bool] =  0
    pid_max =  32768
    threads-max =  1029998
    sem [array] =  250  32000  32  4096
      SEMMSL (max semaphores per array) =  250
      SEMMNS (max sems system-wide)     =  32000
      SEMOPM (max ops per semop call)   =  32
      SEMMNI (max number of sem arrays) =  4096
    shmall [4-KiB pages] =  4294967296  (16384.0 GiB max total shared memory)
    shmmax [bytes] =  68719476736  (64.00 GiB max segment size)
    shmmni [segments] =  4096  (max number of segs)
    sysrq [bitmask] =  0  (disallowed)
  fs.
    file-max [fds] =  6512672  (system-wide limit on nr open file descriptors)
    file-nr [fds] =  2496  0  6512672  (nr allocated fds, N/A, nr free fds)
    inode-nr [inodes] =  78574  1  (nr_inodes allocated, nr_free_inodes)
  net.
    core.rmem_max [bytes] =  131071  (127 KiB)
    core.wmem_max [bytes] =  131071  (127 KiB)
    core.rmem_default [bytes] =  124928  (122 KiB)
    core.wmem_default [bytes] =  124928  (122 KiB)
    ipv4.icmp_echo_ignore_all [bool] =  0
    ipv4.ip_forward [bool] =  0
    ipv4.tcp_max_orphans [sockets] =  65536  (4096 MiB @ max 64 KiB per orphan)
    ipv4.tcp_mem [4-KiB pages] =  6179904  8239872  12359808  (23.57 GiB, 31.43 GiB, 47.15 GiB)
    ipv4.udp_mem [4-KiB pages] =  6179904  8239872  12359808  (23.57 GiB, 31.43 GiB, 47.15 GiB)
    ipv4.tcp_window_scaling [bool] =  1
    ipv4.tcp_rmem [bytes] =  4096  87380  4194304  (4 KiB, 85 KiB, 4096 KiB)
    ipv4.tcp_wmem [bytes] =  4096  16384  4194304  (4 KiB, 16 KiB, 4096 KiB)
    ipv4.udp_rmem_min [bytes] =  4096  (4 KiB)
    ipv4.udp_wmem_min [bytes] =  4096  (4 KiB)
  vm.
    dirty_ratio =  20  (% of total system memory)
    dirty_bytes =  0  (disabled -- check dirty_ratio)
    dirty_background_ratio =  10  (% of total system memory)
    dirty_background_bytes =  0  (disabled -- check dirty_background_ratio)
    dirty_expire_centisecs =  3000
    dirty_writeback_centisecs =  500
    nr_hugepages [2-MiB pages] =  0
    overcommit_memory [0-2] =  0  (heuristic overcommit)
    overcommit_ratio =  50
    oom_kill_allocating_task [bool] =  0  (scan tasklist)
    panic_on_oom [0-2] =  0  (no panic)
    swappiness [0-100] =  60
```


**Finally, here's example of using some of the Special options, which allow you to specify a specific file instead of a full sosreport or having `xsos` run on your local system:**

```
[rsaw]$ xsos -v0 --B /tmp/dmidecode.txt --I=/tmp/ipaddr.dump --P /tmp/psaux.txt
DMIDECODE
  BIOS:
    HP, version P66, 06/24/2011
  System:
    Mfr:  HP
    Prod: ProLiant DL980 G7
    Vers: Not Specified
    Ser:  CZ3XXXXXXXX      
    UUID: 35344D41-XXXXXX-5A43-3331-XXXXXXXXXXX
  CPU:
    8 of 8 CPU sockets populated, 8 cores/16 threads per CPU
    64 total cores, 128 total threads
    Mfr:  Intel
    Fam:  Xeon
    Freq: 2133 MHz
    Vers: Intel(R) Xeon(R) CPU E7- 2830 @ 2.13GHz 
  Memory:
    262144 MB (256 GB) total
    32 of 128 DIMMs populated (system max capacity 512 GB)

IP
  Interface  Slave Of  MAC Address        State  IPv4 Address
  =========  ========  =================  =====  ==================
  lo         -         -                  up     127.0.0.1/8
  lo:0       -         -                  up     192.168.160.20/32
  eth0       -         ZZ:xy:56:01:12:qr  up     192.168.160.11/24
  eth1       -         ZZ:xy:9a:1a:19:qr  up     10.79.4.68/23
  eth1:1     -         ZZ:xy:9a:1a:19:qr  up     10.79.4.70/23

CHANGELOG.md
Changelog for xsos & rsar
===================================================
In descending chronological order with most recent commits on top. See **[commit history page](https://github.com/ryran/xsos/commits/master)** to browse code.

- - -

**NOTE THAT I HAVEN'T TAKEN THE TIME TO KEEP THIS UP TO DATE IN THE LAST 6 MONTHS .... SEE COMMIT HISTORY OR RPM CHANGELOG**

- - -

**2015/01/01 xsos v0.5.5: #90, #108 (new module --softirq/-r), #110**

- Implemented Jamie's [issue 108 - RFE: Alert on network SoftIRQ backlog or budget drops](https://github.com/ryran/xsos/issues/108)
- Updated bash completion file to account for 3 new options (`-r`, `--R`, `--softirq`)
- Implemented Patrick's [issue 90 - RFE: Add tcp_timestamps to sysctl output](https://github.com/ryran/xsos/issues/90)
- Implemeneted [issue 110 - RFE: explicitly require gawk instead of awk](https://github.com/ryran/xsos/issues/110)


**2015/01/01 xsos v0.5.4: fix division by zero error if weird NETDEV input**

- Fixed [issue 109 - awk division by zero error in NETDEV output](https://github.com/ryran/xsos/issues/109)


**2014/09/22 xsos v0.5.3: fix for issue #107**

- Fixed [issue 107 - xsos -k hangs indefinitely](https://github.com/ryran/xsos/issues/107)


**2014/09/15 xsos v0.5.2: fix for issue #106**

- Fixed [issue 106 - xsos -g doesn't handle it gracefully if an ifcfg-bond* file has multiple BONDING_OPTS declarations](https://github.com/ryran/xsos/issues/106)


**2014/09/03 xsos v0.5.1: more enhancements to --ip**

- Closed [issue 104 - xsos --ip --ipv6 improperly handling interfaces with multiple addrs](https://github.com/ryran/xsos/issues/104)
- Fixed [issue 105 - --ip --scrub-ip leaks too much information](https://github.com/ryran/xsos/issues/105) and also changed the character the scrub-ip & scrub-mac options use from `X` to `⣿`


**2014/09/02 xsos v0.5.0: completely reworked --ip output #102**

- Closed [issue 102 - RFE: xsos --ip --ipv6 should tag link-local addresses as such](https://github.com/ryran/xsos/issues/102)
- As part of that, the columns for `--ip` were completely re-arranged and a "Scope" column was added for IPv6


**2014/08/29 xsos v0.4.5: --ip fix for issue #103**

- Closed [issue 103 - xsos --ip problems](https://github.com/ryran/xsos/issues/103)


**2014/07/21 xsos v0.4.4: color fix for issue 100**

- Closed [issue 100 - xsos -t / --mpath doesn't reset color properly when no paths detected](https://github.com/ryran/xsos/issues/100)


**2014/06/22 xsos v0.4.3: rhel7 fixes for --kdump & /proc/cmdline**

- See ongoing [issue 97 - catchall for RHEL7 issues](https://github.com/ryran/xsos/issues/97)


**2014/06/13 xsos v0.4.2: added to bios display; added rhel7 issue support**

- Modified `-b` / `--bios` display to show more info, per [issue 98 - RFE: Add dmidecode "Firmware Revision" to xsos BIOS page](https://github.com/ryran/xsos/issues/98)
- Added RHEL7 to supported systems regex due to GA release a couple days ago (really only affects colorization with `--os`) and tweaked the way `/etc/os-release` is parsed


**2014/05/07 xsos v0.4.1: changed sysctl value color to simple bold**

- Changed sysctl values to use "important" color instead of h3 color (as far as defaults go, that means instead of blue, the values will be just bold)


**2014/05/07 xsos v0.4.0: issue95 & issue96**

- Fixed [issue 95 - taintcheck in xsos -s and xsos -o has problems](https://github.com/ryran/xsos/issues/95)
- Implemented [issue 96 - RFE: xsos --sysctl need moar better](https://github.com/ryran/xsos/issues/96)


**2014/05/02 xsos v0.3.9: quick fix for issue94**

- Fixed [issue 94 - xsos --bonding shows multiple MAC addrs on slave interface](https://github.com/ryran/xsos/issues/94)


**2014/01/10 xsos v0.3.8: speed improvement for --lspci**

- Improved speed of lspci processing


**2014/01/06 xsos v0.3.7: fix for line 2100: syntax error: invalid arithmetic operator**

- Fixed [issue 87 - xsos: line 2100: eth0_12@eth0: syntax error: invalid arithmetic operator (error token is "@eth0")](https://github.com/ryran/xsos/issues/87)


**2013/12/27 xsos v0.3.6: enhancements to ethtool error-parsing**

- See [discussion on ethtool interface error parsing](https://github.com/ryran/xsos/issues/85) for full details
- Added `XSOS_ETHTOOL_ERR_REGEX` env variable (`XSOS_ETHTOOL_ERR_REGEX="(drop|disc|err|fifo|buf|fail|miss|OOB|fcs|full|frags|hdr|tso).*: [^0]"`)
- Added multiple items to above-mentioned regex
- Completely re-tooled the error parsing to gracefully handle `ethtool -S` output with multiple Rx/Tx queues (i.e., if errors are found, the queue heading is printed)
- As a consequence of indentation being used in `ethtool -S` output more and more, xsos no longer replaces all leading whitespace with the `XSOS_INDENT_H2` variable -- instead we just insert the interface name plus a couple spaces and call it good
- Fixed [issue 86 - xsos -n shows same value for RX & TX bytes!](https://github.com/ryran/xsos/issues/86)


**2013/12/17 xsos v0.3.5: added MTU, --scrub-ip & --scrub-mac**

- Added MTU column to `--ip`
- Replaced `--scrub` with `--scrub-ip` & changed `XSOS_SCRUB_OUTPUT` env variable to `XSOS_SCRUB_IP_HN`
- Added `--scrub-mac` and `XSOS_SCRUB_MACADDR `env var to scrub MACs from bonding and ipaddr output
- Updated bash-completion file


**2013/12/16 xsos v0.3.4: multiple env variable changes; ethtool; rhel5 mktemp**

- Colors: `XSOS_COLOR_*` variables now must be defined as simple words -- blue, cyan, red, etc. and BLUE, CYAN, RED, etc. (for bold). Note that users don't need to define these -- they have defaults. But if one wanted to, they could tweak them as environment variables (in e.g. `~/.bash_profile`).
- `XSOS_ALL_VIEW` & `XSOS_DEFAULT_VIEW` environment variables now take simple modules names instead of obfuscated `x=y;` syntax, i.e.: to change the default view seen when xsos is run with no options, you could run `XSOS_DEFAULTVIEW="bios cpu mem" xsos` (or of course set it in a config file as mentioned above).

xsos-bash-completion.bash
# This file is part of xsos, providing intelligent xsos tab-completion for BASH
# Save it to: /etc/bash_completion.d/
#
# Revision date:  2015/06/23 matching up with xsos v0.6.x
# Latest version: <http://github.com/ryran/xsos>
#
# Copyright 2013, 2015 Ryan Sawhill Aroha <rsaw@redhat.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
#    General Public License <gnu.org/licenses/gpl.html> for more details.
#
#-------------------------------------------------------------------------------

_xsos()  {
  
  # Variables
  local curr prev shrtopts longopts
  
  # Wipe out COMPREPLY array
  COMPREPLY=()
  
  # Set cur & prev appropriately
  curr=${COMP_WORDS[COMP_CWORD]}
  prev=${COMP_WORDS[COMP_CWORD-1]}
  
  # Short and long options
  shrtopts="-h -V
            -a -b -o -k -c -f -m -d -t -l -e -r -n -g -i -s -p
            -6 -q -u -v -w
            -x -y -z"
            
  longopts="--help --version
            --all --bios --os --kdump --cpu --intrupt --mem --disks --mpath --lspci --ethtool --softirq --netdev --bonding --ip --net --sysctl --ps
            --B --C --F --M --D --T --L --R --N --G --I --P
            --scrub-ip --scrub-mac --ipv6 --wwid --unit --threads --verbose --width
            --nocolor --less --more"
  
  # Check previous arg to see if we need to do anything special
  case "$prev" in
  
      # Disable autocompletion for solo options that can only be run alone
      -h|--help|-V|--version)
          return 0
          ;;
          
      # These special opts require filenames as arguments
      --B|--C|--F|--M|--D|--T|--L|--R|--N|--G|--I|--P)
          compopt -o plusdirs  # Important!
          COMPREPLY=( $(compgen -f -- "$curr") )
          return 0
          ;;
          
      # For unit, choices are b-t
      -u|--unit)
          COMPREPLY=( $(compgen -W "b k m g t" -- "$curr") )
          return 0
          ;;
          
      # For verbosity, choices are 0-4
      -v|--verbose)
          COMPREPLY=( $(compgen -W "0 1 2 3 4" -- "$curr") )
          return 0
          ;;
          
      # For width, choices are w, 0, or any number
      -w|--width)
          COMPREPLY=( $(compgen -W "w 0" -- "$curr") )
          return 0
  
  esac
  
  # Now that we've made it past the options that require args,
  # we can enable directory completion
  compopt -o plusdirs
   
  if [[ $curr == --* ]]; then
      # If current arg starts w/2 dashes, attempt to autocomplete long opts
      COMPREPLY=( $(compgen -W "$longopts" -- "$curr") )
      return 0
  elif [[ $curr == -* ]]; then
      # Otherwise, if current only starts w/1 dash, attempt autocomplete short opts
      COMPREPLY=( $(compgen -W "$shrtopts" -- "$curr") )
      return 0
  fi
}

# Add the names of any xsos aliases (or alternate file-names) to the end of the following line
# (And remove "x" if you have an "x" command that *isn't* an alias for xsos)
complete -F _xsos xsos x
